# R script: 3-survival_analysis_TCGA-datasets.R
# Author: Inge Seim
# This script will interrogate TCGA data sets and reveal whether your gene(s) of interest are associated with Overall Survival (OS) and/or disease-free survival (DFS).
# To obtain up-to-date OS and DFS information, we queried the cBioPortal for Cancer Genomics (downloaded from http://www.cbioportal.org/study?id=TCGACODE_tcga#clinical --replace 'TCGACODE' with your TCGA data set of interest (e.g. 'prad' for TCGA prostate cancer, PRAD).

# load required library
library(data.table)

# Normalised gene expression values were obtained via the UCSC Xena Browser
#  Reprocessed GTEx, TCGA and TARGET with the same pipeline, removing computational batch effects
#  All 19,131 samples with normalised expression
setwd("~/Downloads/R")

# Load gene expression data
# generated by R script: `2C-input gene signature.R`
geneDF <- readRDS("genesignatureDF.Rds")
geneDF$sample <- row.names(geneDF)
# want the last column, sample, to be first
geneDF <- geneDF[,c(ncol(geneDF),1:(ncol(geneDF)-1))]
#@ head(geneDF)


# describe what TCGA data sets you are interested in
data.set.names <- c("ACC","BLCA","CESC","CHOL","COADREAD","DLBC","ESCA","GBM","HNSC","KICH","KIRC","KIRP","LAML","LGG","LIHC","LUAD","LUSC","MESO","OV","PAAD","PCPG","PRAD","SARC","SKCM","STAD","TGCT","THCA","UCEC","UCS","UVM","BRCA")

# add a description for the output file
data.set.description <- c("Adrenocortical Cancer","Bladder Cancer","Cervical Cancer","Bile Duct Cancer","Colon and Rectal Cancer","Large B-cell Lymphoma","Esophagal Cancer ","Glioblastoma","Head and Neck Cancer","Kidney Chromophobe","Kidney Clear Cell Carcinoma ","Kidney Papillary Cell Carcinoma","Acute Myeloid Leukemia","Lower Grade Glioma ","Liver Cancer","Lung Adenocarcinoma ","Lung Squamous Cell Carcinoma","Mesothelioma","Ovarian Cancer","Pancreatic Cancer","Pheochromocytoma and Paraganglioma","Prostate Cancer","Sarcoma","Melanoma","Stomach Cancer ","Testicular Cancer ","Thyroid Cancer","Endometroid Cancer","Uterine Carcinosarcoma","Ocular Melanomas","Breast Invasive Carcinoma")

# # dummy
# data.set.names <- c("PRAD")
# cancertitle <- "Prostate Cancer"
# numberofdatasets <- 1

# create an output directory (and subdirs)
dir.create("output")
dir.create("output/Robjects")
dir.create("output/survival_curves")
dir.create("output/survival_curves/OS")
dir.create("output/survival_curves/DFS")


# loop-through different data sets here
# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# START OF BIG LOOP
numberofdatasets <- length(data.set.names)
remove(TCGA.analysis.list)

TCGA.analysis.list <- list()
for (i in 1:numberofdatasets)
{
  
  # want to obtain sample name
  temp <- as.list(data.set.names)
  dataset <- as.character(temp[i])
  
  # obtain the cancer's full description
  temp <- as.data.frame(data.set.description)
  cancertitle <- as.character(temp[i,1]) # e.g. Ocular Melanomas
  
  # print FYI
  print(i)
  print(dataset)
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  
  
  # ##############################################################
  # load clinical data
  # ##############################################################
  system.time(clinicalDF <- fread(paste("./data/cBioPortal/",dataset,".txt",sep=""), header=T, sep="\t", stringsAsFactors=T)) # RNA
  clinicalDF <- as.data.frame(clinicalDF)
  names(clinicalDF) <- gsub(" ", ".", names(clinicalDF)) # space incompatible
  names(clinicalDF) <- gsub("\\(", ".", names(clinicalDF)) # space incompatible
  names(clinicalDF) <- gsub("\\)", ".", names(clinicalDF)) # space incompatible
  
  # ==============================================================
  # Some cBioPortal data sets do not denote DFS, but have an equivalent entry we can use.
  # Others have no OS: we will put in dummy values to allow batch processing of TCGA data sets. The dummy will results in obvious 'dummy' survival output
  #
  # FIX OS
  if("Overall.Survival..Months." %in% colnames(clinicalDF))
  {
    cat("Yep, it's in there!\n");
  } else {
    cat("rename!\n");
    names(clinicalDF)[names(clinicalDF) == 'Death.from.Initial.Pathologic.Diagnosis.Date'] <- 'Overall.Survival..Months.'
    clinicalDF$Overall.Survival..Months. <- clinicalDF$Overall.Survival..Months./(365/12)
  }
  # Death.from.Initial.Pathologic.Diagnosis.Date is in days, need to convert to months so all data sets are the same
  colnames(clinicalDF)
  # clinicalDF$Overall.Survival.Status
  # clinicalDF$Overall.Survival..Months.
  
  # FIX DFS
  # 4 have DFS status, but no time points, so ...
  # Disease.Free..Months.  # WITH ALL SET TO 999
  # Disease.Free.Status # WITH ALL SET TO 0
  if("Disease.Free..Months." %in% colnames(clinicalDF))
  {
    cat("Yep, it's in there!\n");
  } else {
    cat("add dummy!\n");
    sizeofclinicalDF <- length(clinicalDF$Sample.ID)
    dummyDFSmonth <- rep(0, each=sizeofclinicalDF)
    # dummyDFSstatus <- rep(0, each=sizeofclinicalDF) # will rename later
    dummyDFSstatus <- rep("DiseaseFree", each=sizeofclinicalDF)
    clinicalDF$Disease.Free..Months. <- dummyDFSmonth
    clinicalDF$Disease.Free.Status <- dummyDFSstatus
  }
  colnames(clinicalDF)
  clinicalDF$Disease.Free..Months.
  clinicalDF$Disease.Free.Status
  # END OF OS, DFS 'fix'
  # ==============================================================
  # only select a few, common data columns
  clinicalDF <- subset(clinicalDF, select=c(Sample.ID,Sample.Type,Disease.Free..Months.,Disease.Free.Status,Overall.Survival..Months.,Overall.Survival.Status))
  
  # rename columns to something more conventional
  names(clinicalDF)[names(clinicalDF) == 'Sample.ID'] <- 'sample'
  names(clinicalDF)[names(clinicalDF) == 'Sample.Type'] <- 'sample_type'
  names(clinicalDF)[names(clinicalDF) == 'Overall.Survival.Status'] <- 'OS_EVENT'
  names(clinicalDF)[names(clinicalDF) == 'Overall.Survival..Months.'] <- 'OS'
  names(clinicalDF)[names(clinicalDF) == 'Disease.Free..Months.'] <- 'DFS'
  names(clinicalDF)[names(clinicalDF) == 'Disease.Free.Status'] <- 'DFS_EVENT'
  
  # get rid of Solid Normals. Most data sets have very few and we are interested in cancer outcomes
  clinicalDF <- (clinicalDF[grep("*Solid*", clinicalDF$sample_type, invert=TRUE), ]) # which are not Solid Normal
  # length(which(clinicalDF$DFS_EVENT =="Recurred/Progressed")) # check no. of DFS
  # length(which(clinicalDF$OS_EVENT =="DECEASED")) # check no. of OS
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # end of clinical data
  # ##############################################################
  
  
  # ##############################################################
  # merge the data
  # ##############################################################
  mergedDF <- merge(clinicalDF,geneDF, by.x='sample', by.y='sample',all=FALSE) # all=FALSE ... do not keep
  

  # ##############################################################
  # remove uninformative data points
  # ##############################################################
  # only keep samples with an OS EVENT
  which(mergedDF$OS_EVENT == "LIVING" | mergedDF$OS_EVENT == "DECEASED")
  # kills one
  mergedDF <- mergedDF[which(mergedDF$OS_EVENT == "LIVING" | mergedDF$OS_EVENT == "DECEASED"),]
  # only keep samples with a DFS EVENT
  which(mergedDF$DFS_EVENT == "Recurred/Progressed" | mergedDF$DFS_EVENT == "DiseaseFree")
  # leaves 464 (minus any OS)  
  mergedDF <- mergedDF[which(mergedDF$DFS_EVENT == "Recurred/Progressed" | mergedDF$DFS_EVENT == "DiseaseFree"),]
  
  
  
  
  
  
  
  
  
  
  # ##############################################################
  # k-means clustering
  # ##############################################################
  GenesOnly <- length(clinicalDF)+1 # only take the means of the genes
  #
  set.seed(20)
  k <- 2 # we want *k* groups to be 2
  numberofruns <- 500
  samplecluster <- kmeans(mergedDF[, GenesOnly:ncol(mergedDF)], k, nstart = numberofruns)
  
  # mergedDF[, GenesOnly:ncol(mergedDF)]
  
  # assign
  samplecluster$cluster <- as.factor(samplecluster$cluster)
  
  # ADD THE GROUP INFORMATION TO YOUR ORIGINAL DATA FRAME
  # append cluster assignment
  mergedDF <- data.frame(mergedDF, samplecluster$cluster)
  # rename to fit preferred style
  names(mergedDF)[names(mergedDF) == 'samplecluster.cluster'] <- 'group'
  
  # assign a more useful name
  levels(mergedDF$group)[levels(mergedDF$group)=="1"] <- "cluster1"
  levels(mergedDF$group)[levels(mergedDF$group)=="2"] <- "cluster2"
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  
  
  # ##############################################################
  # survival analysis
  # ##############################################################
  library(survival)
  library(ggplot2)
  library(ggthemes)
  library(survminer) # required for ggsurvplot
  
  tempdata <- mergedDF
  # omit rows with DFS or OS <NA>
  # DFS
  # class(tempdata$DFS_EVENT) # factor
  tempdata$DFS_EVENT <- as.character(tempdata$DFS_EVENT)
  tempdata <- tempdata[!is.na(tempdata$DFS_EVENT),]
  tempdata[tempdata=="DiseaseFree"]<-0
  tempdata[tempdata=="Recurred/Progressed"]<-1
  tempdata$DFS_EVENT <- as.numeric(tempdata$DFS_EVENT)
  #
  # OS
  # class(tempdata$OS_EVENT) # factor
  tempdata$OS_EVENT <- as.character(tempdata$OS_EVENT)
  tempdata <- tempdata[!is.na(tempdata$OS_EVENT),]
  tempdata[tempdata=="LIVING"]<-0
  tempdata[tempdata=="DECEASED"]<-1
  tempdata$OS_EVENT <- as.numeric(tempdata$OS_EVENT)
  
  # save for later perusal
  save(tempdata,file=paste("./", "output/Robjects","/",dataset,"ready",".Robj",sep="")) #
  
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # DFS
  survdiff(Surv(DFS, DFS_EVENT) ~ group ,data=tempdata , rho=0)  # save
  DFSsurv <- survdiff(Surv(DFS, DFS_EVENT) ~ group ,data=tempdata, rho=0)
  
  p.val.DFS <- 1 - pchisq(DFSsurv$chisq, length(DFSsurv$n) - 1)
  
  data.set.size <- length((tempdata$group)) # total number of samples = 489 in PRAD
  data.set.DFS <- length(which(tempdata$DFS_EVENT==1)) # 
  
  cluster1.size <- length(which(tempdata$group=="cluster1" )) # all cluster1 samples = 157 in PRAD
  cluster1.DFS <- length(which(tempdata$DFS_EVENT==1 & tempdata$group=="cluster1" )) # all cluster1 DFS relapsed = 43 in PRAD
  cluster2.size <- length(which(tempdata$group=="cluster2" )) # all cluster2 samples = 332 in PRAD
  cluster2.DFS <- length(which(tempdata$DFS_EVENT==1 & tempdata$group=="cluster2" )) # all cluster2 DFS relapsed = 48 in PRAD
  
  # cluster 1
  temp <- tempdata
  temp <- temp[(which(temp$group=="cluster1" & temp$DFS_EVENT==1)),]  # only with relapse!
  # median, mean±SEM
  #@ tempdata$DFS # in *months*
  cluster1.median.DFS <- (median(temp$DFS,na.rm = TRUE))/12*365 # *days*
  #
  cluster1.mean.DFS <- (mean(temp$DFS,na.rm = TRUE))/12*365 # *days*
  # STANDARD ERROR OF THE *MEAN*
  sem <- function(x) sqrt(var(x,na.rm=TRUE)/length(na.omit(x))) # remove pesky NAs!
  cluster1.sem.DFS <- sem(temp$DFS)/12*365
  # cluster 2
  temp <- tempdata
  temp <- temp[(which(temp$group=="cluster2" & temp$DFS_EVENT==1)),]  # only with relapse!
  # median, mean±SEM
  #@ tempdata$DFS # in *months*
  cluster2.median.DFS <- (median(temp$DFS,na.rm = TRUE))/12*365 # *days*
  #
  cluster2.mean.DFS <- (mean(temp$DFS,na.rm = TRUE))/12*365 # *days*
  # STANDARD ERROR OF THE *MEAN*
  sem <- function(x) sqrt(var(x,na.rm=TRUE)/length(na.omit(x))) # remove pesky NAs!
  cluster2.sem.DFS <- sem(temp$DFS)/12*365
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # OS
  survdiff(Surv(OS, OS_EVENT) ~ group ,data=tempdata, rho=0)  # save
  OSsurv <- survdiff(Surv(OS, OS_EVENT) ~ group ,data=tempdata, rho=0)
  #
  p.val.OS <- 1 - pchisq(OSsurv$chisq, length(OSsurv$n) - 1)
  #
  data.set.size <- length((tempdata$group)) # total number of samples = 489 in PRAD
  data.set.OS <- length(which(tempdata$OS_EVENT==1)) # 
  #
  cluster1.size <- length(which(tempdata$group=="cluster1" )) # all cluster1 samples = 157 in PRAD
  cluster1.OS <- length(which(tempdata$OS_EVENT==1 & tempdata$group=="cluster1" )) # 
  cluster2.size <- length(which(tempdata$group=="cluster2" )) # all cluster2 samples = 332 in PRAD
  cluster2.OS <- length(which(tempdata$OS_EVENT==1 & tempdata$group=="cluster2" )) # 
  #
  # cluster 1
  temp <- tempdata
  temp <- temp[(which(temp$group=="cluster1" & temp$OS_EVENT==1)),]  # only with relapse!
  # median, mean±SEM
  # tempdata$OS # in *months*
  cluster1.median.OS <- (median(temp$OS,na.rm = TRUE))/12*365 # *days*
  #
  cluster1.mean.OS <- (mean(temp$OS,na.rm = TRUE))/12*365 # *days*
  # STANDARD ERROR OF THE *MEAN*
  sem <- function(x) sqrt(var(x,na.rm=TRUE)/length(na.omit(x))) # remove pesky NAs!
  cluster1.sem.OS <- sem(temp$OS)/12*365
  # cluster 2
  temp <- tempdata
  temp <- temp[(which(temp$group=="cluster2" & temp$OS_EVENT==1)),]  # only with relapse!
  # median, mean±SEM
  # tempdata$OS # in *months*
  cluster2.median.OS <- (median(temp$OS,na.rm = TRUE))/12*365 # *days*
  #
  cluster2.mean.OS <- (mean(temp$OS,na.rm = TRUE))/12*365 # *days*
  # STANDARD ERROR OF THE *MEAN*
  sem <- function(x) sqrt(var(x,na.rm=TRUE)/length(na.omit(x))) # remove pesky NAs!
  cluster2.sem.OS <- sem(temp$OS)/12*365
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  
  # can now save all important information
  # add % DFS OR %OS column
  # Note: a bit crude, as many papers deals with e.g. 5-year survival, but gives an initial estimate.
  # S as the estimated probability of surviving to time t for those alive just before t multiplied by the proportion of subjects surviving to t.
  # S = 100% - ((n DFS cluster X/n total cluster X)*100)%
  cluster1.DFS.rate <- 100-(   ceiling((cluster1.DFS/cluster1.size)*100)   )
  cluster2.DFS.rate <- 100-( ceiling(cluster2.DFS/cluster2.size*100)   )
  cluster1.OS.rate <- 100-( ceiling(cluster1.OS/cluster1.size*100)   )
  cluster2.OS.rate <- 100-( ceiling(cluster2.OS/cluster2.size*100)   )
  # @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
  
  # join them
  dataDF <- as.data.frame(cbind(dataset,cancertitle, data.set.size, data.set.DFS,cluster1.size, cluster1.DFS, cluster2.size, cluster2.DFS,cluster1.DFS.rate,cluster2.DFS.rate,cluster1.median.DFS, cluster1.mean.DFS, cluster1.sem.DFS,cluster2.median.DFS, cluster2.mean.DFS, cluster2.sem.DFS,p.val.DFS, data.set.size, data.set.OS, cluster1.size, cluster1.OS, cluster2.size, cluster2.OS,cluster1.OS.rate,cluster2.OS.rate,cluster1.median.OS, cluster1.mean.OS, cluster1.sem.OS,cluster2.median.OS, cluster2.mean.OS, cluster2.sem.OS,p.val.OS))
  
  colnames(dataDF) <- c("TCGA.data.set", "TCGA.name","no.samples", "no.DFS.total","no.cluster1", "no.cluster1.DFS", "no.cluster2", "no.cluster2.DFS","DFS.rate.cluster1","DFS.rate.cluster2","median.DFS.cluster1", "mean.DFS.cluster1", "sem.DFS.cluster1","median.DFS.cluster2", "mean.DFS.cluster2", "sem.DFS.cluster2","p.val.DFS","no.samples", "no.OS.total","no.cluster1", "no.cluster1.OS", "no.cluster2", "no.cluster2.OS","OS.rate.cluster1","OS.rate.cluster2","median.OS.cluster1", "mean.OS.cluster1", "sem.OS.cluster1","median.OS.cluster2", "mean.OS.cluster2", "sem.OS.cluster2","p.val.OS")
  # head(dataDF)
  #
  TCGA.analysis.list[[i]] <- dataDF  # save for each i (here: TCGA data set)
  # save data for all
  print(i)
  
  # ##############################################################
  # DRAW SURVIVAL CURVES
  # ##############################################################
  tempdataYear <- tempdata
  tempdataYear$DFS # months
  tempdataYear$OS # months
  tempdataYear$DFS <- tempdataYear$DFS/12 # years
  tempdataYear$OS <- tempdataYear$OS/12 # years
  #
  # DFS
  #@ survdiff(Surv(OS, DFS_EVENT) ~ group ,data=tempdata, rho=0)  # save
  fit=survfit(Surv(DFS, DFS_EVENT) ~ group,data=tempdataYear)
  # desired colours
  # #f42732 = red
  # #000000 = black
  # https://cran.r-project.org/web/packages/survminer/survminer.pdf
  pdf(paste("./output/survival_curves/DFS/",dataset,"-DFS.pdf",sep=""))
  p <- ggsurvplot(fit, risk.table = TRUE,
                  pval = TRUE, risk.table.y.text.col = TRUE, break.time.by=1,conf.int=FALSE, censor=TRUE, legend="top", palette=c("#f42732","#000000"))
  p$table <- p$table
  print(p)
  dev.off()
  
  # OS
  #@ survdiff(Surv(OS, OS_EVENT) ~ group ,data=tempdata, rho=0)  # save
  fit=survfit(Surv(OS, OS_EVENT) ~ group,data=tempdataYear)
  # desired colours
  # #f42732 = red
  # #000000 = black
  # https://cran.r-project.org/web/packages/survminer/survminer.pdf
  pdf(paste("./output/survival_curves/OS/",dataset,"-OS.pdf",sep=""))
  p <- ggsurvplot(fit, risk.table = TRUE,
                  pval = TRUE, risk.table.y.text.col = TRUE, break.time.by=1,conf.int=FALSE, censor=TRUE, legend="top", palette=c("#f42732","#000000"))
  p$table <- p$table
  print(p)
  dev.off()
  # size.est= 1.0 size of the lines
  
  
} # END OF BIG LOOP
# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++


# ##############################################################
# SAVE THE OUTPUT
# ##############################################################
TCGAoutputDF <- do.call(rbind.data.frame, TCGA.analysis.list)  # convert the list to a data frame!
rownames(TCGAoutputDF) <- TCGA.analysis.list$TCGAoutputDF
head(TCGAoutputDF)

# DFS
# colnames(TCGAoutputDF)
DFS.TCGAoutputDF <- TCGAoutputDF[,1:17]
OS.TCGAoutputDF <- cbind(TCGAoutputDF[,1:2],TCGAoutputDF[,18:32])

# save outputs
save(DFS.TCGAoutputDF,file = "./output/DFS.gene-signature-vs-TCGA.Robj")  #
save(OS.TCGAoutputDF,file = "./output/OS.gene-signature-vs-TCGA.Robj")  #

# save as a TSV
write.table(DFS.TCGAoutputDF, file='./output/DFS.gene-signature-vs-TCGA.tsv', quote=FALSE, sep='\t', col.names = TRUE)
write.table(OS.TCGAoutputDF, file='./output/OS.gene-signature-vs-TCGA.tsv', quote=FALSE, sep='\t', col.names = TRUE)
# ~~~~~~~~~~~~~~~~~~~~~~~~ THIS IS THE END ~~~~~~~~~~~~~~~~~~~~~~~~ #
